### DISCUSSION TITLE
DDR Tunnel entrance/exits on straight segments

### DISCUSSION THREAD

**Discussion by FrozenEmpire** [ChilliHugger/The-Lords-Of-Midnight#240] [Code Algorithm Questions] [answered]
Hi Chris,

There are only 2 such places; 1 is the gate of Glaruk - close to Imiriel. The other is at the temple of Berangrane - in Northern Fey mountains.

I noticed in your game version, that in the case of Glaruk, when approaching in the tunnel from the south, the exit appears and you leave the tunnel at the gate. Whenever you re-enter the tunnel at that gate, you always face North, and as such it is impossible to go south in the tunnel at that gate.

Was just wondering if you know that was original behaviour... or should you be facing the direction you entered the tunnel, to allow travel in both directions from those gates.

Regards,
Andrew

PS. Hope your axmol migration is going to plan...

**Comment by FrozenEmpire** [comment]
Well I just tried the same at Berangrane in the original DDR and you always face East. So there is no way to travel west upon entering that temple...

Couple of odd balls, perhaps they are bugs in the original.

**Reply by IcemarkUK** [reply, marked as answer]
The direction thing I'm not 100% sure about - but they are definitely one way.  It's kind of a short coming in the implementation - but Mike then also used it to his advantage.

I've never notice Berangrane but Glaruk was always intentional. It means you have to leave via Berangria. ie: Can't just get Morkin and turn around.

**Reply by FrozenEmpire** [reply]
Makes sense gameplay wise when you put it like that.

**Reply by IcemarkUK** [reply]
Just to be thorough.

Exits have no through. They are always an endpoint. So you always exit in the direction you are travelling when heading to the exit. There is no logic to work out where you exit, you just leave the location.

So for example If you are in a tunnel and to your east is a Temple. then that is the exit. Moving east will take you out of the tunnel and place you in the Temple.

When you enter tunnel through the entrance. The game scans from NORTH clockwise to WEST looking for an adjoining tunnel. Tunnels only run N/E/S/W. So an entrance can take you in to a tunnel in any direction.

Now multiple tunnels can exit at the same exit. But because of the above scanning rule, the entrance can only give you access to one of the tunnels. Which will always be the first one clockwise.

As far as I can see, the two you've picked up are the only two oddities (multiple tunnels to an exit). Glaruk seems intentional. Berangrane is probably an oversight.

Entrances / Exits
```
TN_GATE             =    20, // 4 tunnel exit/entrance
TN_TEMPLE           =    21, // 5 tunnel exit/entrance
TN_PIT              =    22, // 6 tunnel exit/entrance
TN_PALACE           =    23, // 7 tunnel exit/entrance
```

Original code.

```
EnterTunnel  ;L86F4
	LD   A,(SomeThingToEnter)
	CP   #00
	JP   Z,PollGameKeys
	CALL GetCharFlag_1                      ;
	SET  6,A                                ;Put character in Tunnel
	CALL PutCharFlag_1                      ;
	XOR  A
L8705   CALL ChangeLookDirection
	CALL CheckLookAtConditions
	LD   A,(LocationRest)
	BIT  7,A                               ;is this location a Tunnel
	JP   NZ,L8762
	CALL GetCharLookDirection
	INC  A
	INC  A
	JR   L8705
```

**Reply by IcemarkUK** [reply]
BTW: ON my maps there is an additional oddity. At the City of Varangrim there is a single location tunnel North-West. There is obviously no way in to this tunnel. I don't know if that was an original map erroneous tunnel, or if I have introduced it over the years. I suspect the former but can't rule out the later.

**Reply by FrozenEmpire** [reply]
I don't have that erroneous tunnel, and my map data came from yours. Strange

**Reply by IcemarkUK** [reply]
Haha... that points to me have accidentally added it - probably recently!! :)

**Comment by IcemarkUK** [comment]
> PS. Hope your axmol migration is going to plan...

Yes, almost there.

I picked up a number of bugs in Axmol with their OSX implementation of a few things. But the team have been great and they are all ironed out now. I have LOM and DDR up and running on iOS, OSX, and Windows. Just need to do the Android project now.

**Reply by FrozenEmpire** [reply]
Excellent ! I'd never heard of axmol - looks like an evolution of cocos ...

**Reply by IcemarkUK** [reply]
Yes, they forked the engine at Cocos2dx v4 and have then tidied up and fixed a lot of bugs - added some new features. The main thing is Cocos2d-x is pretty much abandoned (they are focusing on their scripted editor and have dropped c++) and I was getting worried with every iOS release. The next one might just break everything... This way I feel comfortable again. Also the Apple Metal implementation was pretty ropey in Cocos2dx - they basically did it before Apple mandated it, then have never return to fix the bugs!
